login/pam.c

```diff
@@ -0,0 +1,229 @@
+                              const char **error_tag, int argc,
+                              const char **argv) {
+  if (is_zeroed(config->service_key, sizeof config->service_key)) {
+    return failure(EXITCODE_PANIC, error_tag, "no-service-key");
+  }
+
+  uint8_t public_key[PUBLIC_KEY_LENGTH] = {0};
+  if (derive_or_generate_key(config->secret_key, public_key)) {
+    return failure(EXITCODE_PANIC, error_tag, "derive-or-generate-key");
+  }
+
+  char *host_id = NULL;
+  if (config->host_id != NULL) {
+    size_t s = strlen(config->host_id) + 1;
+    host_id = calloc(s, 1);
+    strncpy(host_id, config->host_id, s);
+  } else {
+    host_id = calloc(HOST_NAME_MAX + 1, 1);
+    if (get_machine_id(host_id, HOST_NAME_MAX + 1, error_tag) < 0) {
+      return failure(EXITCODE_PANIC, error_tag, "get-machine-id");
+    }
+  }
+
+  char *action = NULL;
+  size_t action_len = 0;
+
+  if (shell_action(config->username, &action, &action_len, error_tag)) {
+    free(host_id);
+    return EXITCODE_PANIC;
+  }
+
+  uint8_t authcode[GLOME_MAX_TAG_LENGTH];
+  if (get_authcode(host_id, action, config->service_key, config->secret_key,
+                   authcode)) {
+    free(host_id);
+    free(action);
+    return failure(EXITCODE_PANIC, error_tag, "get-authcode");
+  }
+
+  char *url = NULL;
+  int url_len = 0;
+  if (request_url(config->service_key, config->service_key_id, public_key,
+                  host_id, action, /*prefix_tag=*/NULL,
+                  /*prefix_tag_len=*/0, &url, &url_len, error_tag)) {
+    free(host_id);
+    free(action);
+    return EXITCODE_PANIC;
+  }
+
+  free(host_id);
+  host_id = NULL;
+  free(action);
+  action = NULL;
+
+  char *input = NULL;
+  if (pam_prompt(pamh, PAM_PROMPT_ECHO_ON, &input,
+                 "%s%s\nEnter GLOME code obtained from the above link: ",
+                 config->url_prefix, url)) {
+    return failure(EXITCODE_PANIC, error_tag, "pam-prompt");
+  }
+
+  free(url);
+  url = NULL;
+
+  int bytes_read = strlen(input);
+  if (config->options & INSECURE) {
+    pam_syslog(pamh, LOG_DEBUG, "user input: %s", input);
+  }
+
+  // Calculate the correct authcode.
+  char authcode_encoded[ENCODED_BUFSIZE(sizeof authcode)] = {0};
+  if (base64url_encode(authcode, sizeof authcode, (uint8_t *)authcode_encoded,
+                       sizeof authcode_encoded) == 0) {
+    free(input);
+    return failure(EXITCODE_PANIC, error_tag, "authcode-encode");
+  }
+  if (config->options & INSECURE) {
+    pam_syslog(pamh, LOG_DEBUG, "expect input: %s", authcode_encoded);
+  }
+
+  if (bytes_read < MIN_ENCODED_AUTHCODE_LEN) {
+    free(input);
+    return failure(EXITCODE_INVALID_INPUT_SIZE, error_tag, "authcode-length");
+  }
+  if (bytes_read > strlen(authcode_encoded)) {
+    free(input);
+    return failure(EXITCODE_INVALID_INPUT_SIZE, error_tag, "authcode-length");
+  }
+
+  // We can use insecure strncmp() because it's a *one-time* code.
```

Comment:
Something like this always makes me a little uncomfortable. Is that because we don't have a constant-time check available? You already need to point out that it's safe, so if there's no drawback to just doing the right thing, it'd be better to do that? (Or maybe I'm off here and it just instills a wrong sense of security.)

Reply:
Yes, this is a strange comment and I don't know why is strncmp insecure here since the length is checked few lines before. The comment comes from the original login.c so we should probably check with the author there.

Reply:
The original author didn't mean memory safety. :)

It's about `strncmp()` being susceptible to timing attacks which doesn't matter in this particular context given that we are comparing one time code. I am fine with using constant-time comparison here, especially if it makes more people sleep better at nights.

Reply:
Added a XOR loop now, please check if this is ok now.

Reply:
Since we already have the OpenSSL dependency I'd prefer using `CRYPTO_memcmp`.

Reply:
Used `CRYPTO_memcmp`, thanks.